core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java

```diff
@@ -226,6 +241,17 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
         return selectDistributionNodes(nodeMap.get().get(), nodeTaskMap, maxSplitsWeightPerNode, maxPendingSplitsWeightPerTask, maxUnacknowledgedSplitsPerTask, splits, existingTasks, bucketNodeMap);
     }
 
+    private List<InternalNode> getFreeNodesForStage(NodeAssignmentStats assignmentStats, List<InternalNode> nodes)
+    {
+        ImmutableList.Builder<InternalNode> freeNodes = ImmutableList.builder();
+        for (InternalNode node : nodes) {
+            if (assignmentStats.getTotalSplitsWeight(node) < maxSplitsWeightPerNode && assignmentStats.getUnacknowledgedSplitCountForStage(node) < maxUnacknowledgedSplitsPerTask) {
+                freeNodes.add(node);
+            }
+        }
+        return freeNodes.build();
```

Comment:
nit: you can use stream for that.

Reply:
This is in relatively tight loop, so it's better to avoid streams